codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +244,41 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
+                // Write the initial headers on the stream immediately
+                stream.open(endOfStream);
+                future = frameWriter.writeHeaders(ctx, stream.id(), headers, streamDependency, weight,
+                        exclusive, padding, endOfStream, promise);
+                ctx.flush();
+            } else {
+                // Pass intermediate & trailing headers to the flow-controller so it can maintain their
+                // sequence relative to DATA frames.
+                future = flowController().sendTrailers(ctx, stream, headers, padding, endOfStream, promise);
             }
+            if (endOfStream) {
+                // Flag delivery of EOS synchronously to prevent subsequent frames being enqueued in the flow
+                // controller.
+                stream.endOfStreamSent();
+            }
+            future.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    if (!future.isSuccess()) {
+                        // The write failed, handle the error.
+                        lifecycleManager.onException(ctx, future.cause());
+                    } else if (endOfStream) {
+                        lifecycleManager.closeLocalSide(stream, ctx.newPromise());
```

Comment:
Depending on how we resolve https://github.com/netty/netty/pull/3353/files#r23606075...we may want to extract this out for the case when we call `writeHeaders` directly.